File: ListaEnlazada/codigo_python/lista_enlazada.py
"""
CONCEPTO: LISTAS ENLAZADAS EN PYTHON
En Python no necesitamos gestionar la memoria manualmente (no hay malloc ni free).
El recolector de basura (Garbage Collector) se encarga de liberar los objetos
cuando ya no hay variables apuntando a ellos.
"""

class Nodo:
    """
    Clase que representa un nodo de la lista enlazada.
    """
    def __init__(self, dato):
        self.dato = dato          # El valor entero (u otro tipo)
        self.siguiente = None     # Referencia al siguiente nodo (nuestro "apuntador")

class ListaEnlazada:
    def __init__(self):
        self.cabeza = None        # La lista inicia vacía

    def insertar_al_principio(self, valor):
        # 1. Creamos el nuevo nodo
        nuevo_nodo = Nodo(valor)
        # 2. Conectamos el nuevo nodo con el que era la cabeza actual
        nuevo_nodo.siguiente = self.cabeza
        # 3. La nueva cabeza es el nuevo nodo
        self.cabeza = nuevo_nodo
        print(f"Nodo con valor {valor} insertado.")

    def buscar(self, valor_buscado):
        actual = self.cabeza
        # Recorremos la lista mientras no lleguemos al final
        while actual is not None:
            if actual.dato == valor_buscado:
                return actual     # Retorna la referencia al objeto Nodo
            actual = actual.siguiente
        return None

    def modificar(self, valor_buscado, nuevo_valor):
        nodo_encontrado = self.buscar(valor_buscado)
        if nodo_encontrado is not None:
            nodo_encontrado.dato = nuevo_valor
            print(f"Valor {valor_buscado} modificado por {nuevo_valor} con éxito.")
        else:
            print(f"Error: No se encontró el valor {valor_buscado}.")

    def eliminar(self, valor_eliminar):
        temp = self.cabeza
        previo = None
        
        # Caso 1: El nodo a eliminar es la cabeza
        if temp is not None and temp.dato == valor_eliminar:
            self.cabeza = temp.siguiente
            # En Python, simplemente perdiendo la referencia al objeto original,
            # el recolector de basura lo eliminará de la RAM.
            print(f"Nodo {valor_eliminar} eliminado (era la cabeza).")
            return

        # Caso 2: Buscar en el resto de la lista
        while temp is not None and temp.dato != valor_eliminar:
            previo = temp
            temp = temp.siguiente
            
        # Si no se encontró
        if temp is None:
            print(f"Error: Nodo {valor_eliminar} no encontrado.")
            return
            
        # Desvinculamos el nodo
        previo.siguiente = temp.siguiente
        print(f"Nodo {valor_eliminar} eliminado de la lista.")

    def imprimir_lista(self):
        print("Estado de la lista: ", end="")
        actual = self.cabeza
        while actual is not None:
            print(f"{actual.dato} -> ", end="")
            actual = actual.siguiente
        print("None\n")


# Bloque de pruebas
if __name__ == "__main__":
    lista = ListaEnlazada()
    
    lista.insertar_al_principio(10)
    lista.insertar_al_principio(20)
    lista.insertar_al_principio(30)
    lista.insertar_al_principio(40)
    lista.imprimir_lista()
    
    resultado = lista.buscar(20)
    if resultado:
        print(f"Búsqueda: Se encontró el nodo con valor {resultado.dato}.\n")
        
    lista.modificar(30, 35)
    lista.imprimir_lista()
    
    lista.eliminar(40)
    lista.imprimir_lista()
    
    lista.eliminar(10)
    lista.imprimir_lista()

File: Ordenamiento/ListaEnlazada/codigo_c/lista_enlazada.c
#include <stdio.h>
#include <stdlib.h>

// Definición de la estructura del nodo para la lista enlazada
struct Nodo {
    int dato;              // El valor entero que almacena el nodo
    struct Nodo* siguiente; // Apuntador al siguiente nodo en la lista
};

// Función para crear un nuevo nodo
// Recibe el valor entero que almacenará el nodo
struct Nodo* crearNodo(int valor) {
    // Se reserva memoria dinámica para el nuevo nodo usando malloc
    struct Nodo* nuevoNodo = (struct Nodo*)malloc(sizeof(struct Nodo));
    
    // Si malloc falla (devuelve NULL), significa que no hay memoria disponible
    if (nuevoNodo == NULL) {
        printf("Error: No se pudo asignar memoria para el nuevo nodo.\n");
        exit(1); // Se termina la ejecución del programa con código de error
    }
    
    // Se asigna el valor al campo 'dato' del nuevo nodo
    nuevoNodo->dato = valor;
    // Como el nodo se insertará al final o es nuevo, su apuntador 'siguiente' es NULL
    nuevoNodo->siguiente = NULL;
    
    // Se retorna el apuntador al nodo recién creado
    return nuevoNodo;
}

// Función para insertar un nodo al principio de la lista
// Recibe un apuntador al apuntador de la cabeza (para poder modificar la cabeza real) y el valor
void insertarAlPrincipio(struct Nodo** cabeza_ref, int valor) {
    // Se crea el nuevo nodo llamando a la función crearNodo
    struct Nodo* nuevoNodo = crearNodo(valor);
    
    // El 'siguiente' del nuevo nodo apuntará al nodo que actualmente es la cabeza
    nuevoNodo->siguiente = *cabeza_ref;
    
    // Ahora la nueva cabeza de la lista será el nuevo nodo
    *cabeza_ref = nuevoNodo;
    
    printf("Nodo con valor %d insertado en la lista.\n", valor);
}

// Función para buscar un nodo por su valor
// Retorna un apuntador al nodo si lo encuentra, o NULL si no lo encuentra
struct Nodo* buscar(struct Nodo* cabeza, int valorBuscado) {
    // Se usa un apuntador temporal 'actual' para recorrer la lista, empezando por la cabeza
    struct Nodo* actual = cabeza;
    
    // Mientras no lleguemos al final de la lista (actual no sea NULL)
    while (actual != NULL) {
        // Si el dato del nodo actual coincide con el valor buscado
        if (actual->dato == valorBuscado) {
            return actual; // Retornamos el apuntador a este nodo
        }
        // Si no es el valor, avanzamos al siguiente nodo de la lista
        actual = actual->siguiente;
    }
    
    // Si el bucle termina y no se encontró el valor, se retorna NULL
    return NULL;
}

// Función para modificar el valor de un nodo existente
// Recibe la cabeza de la lista, el valor a buscar y el nuevo valor
void modificar(struct Nodo* cabeza, int valorBuscado, int nuevoValor) {
    // Reutilizamos la función 'buscar' para encontrar el nodo a modificar
    struct Nodo* nodoEncontrado = buscar(cabeza, valorBuscado);
    
    // Si la función buscar no retornó NULL, significa que el nodo existe
    if (nodoEncontrado != NULL) {
        // Se actualiza el campo 'dato' del nodo encontrado con el nuevo valor
        nodoEncontrado->dato = nuevoValor;
        printf("Valor %d modificado por %d con exito.\n", valorBuscado, nuevoValor);
    } else {
        // Si retornó NULL, el nodo no existe en la lista
        printf("Error: No se encontro un nodo con el valor %d para modificar.\n", valorBuscado);
    }
}

// Función para eliminar un nodo dado su valor
// Recibe el apuntador al apuntador de la cabeza y el valor a eliminar
void eliminar(struct Nodo** cabeza_ref, int valorEliminar) {
    // Apuntador 'temp' para recorrer la lista, inicializado en la cabeza
    struct Nodo* temp = *cabeza_ref;
    // Apuntador 'previo' para mantener la referencia al nodo anterior a 'temp'
    struct Nodo* previo = NULL;
    
    // Caso 1: El nodo a eliminar es la cabeza de la lista
    if (temp != NULL && temp->dato == valorEliminar) {
        // La nueva cabeza será el siguiente nodo del actual
        *cabeza_ref = temp->siguiente;
        // Se libera la memoria del nodo eliminado
        free(temp);
        printf("Nodo con valor %d eliminado (era la cabeza).\n", valorEliminar);
        return; // Terminamos la función
    }
    
    // Caso 2: El nodo a eliminar está en medio o al final
    // Recorremos la lista buscando el valor, llevando registro del nodo anterior
    while (temp != NULL && temp->dato != valorEliminar) {
        previo = temp;          // El nodo actual se convierte en el previo
        temp = temp->siguiente; // Avanzamos al siguiente nodo
    }
    
    // Si 'temp' llegó a NULL, significa que recorrimos toda la lista y no encontramos el valor
    if (temp == NULL) {
        printf("Error: No se encontro el nodo con valor %d para eliminar.\n", valorEliminar);
        return;
    }
    
    // Si se encontró, el 'siguiente' del nodo previo ahora apuntará al 'siguiente' del nodo a eliminar
    // Desvinculando así el nodo 'temp' de la lista
    previo->siguiente = temp->siguiente;
    
    // Se libera la memoria ocupada por el nodo 'temp'
    free(temp);
    printf("Nodo con valor %d eliminado de la lista.\n", valorEliminar);
}

// Función auxiliar para imprimir la lista (útil para verificar el funcionamiento)
void imprimirLista(struct Nodo* nodo) {
    printf("Estado de la lista: ");
    // Recorremos la lista hasta el final
    while (nodo != NULL) {
        // Imprimimos el dato del nodo actual
        printf("%d -> ", nodo->dato);
        // Avanzamos al siguiente nodo
        nodo = nodo->siguiente;
    }
    // Indicamos el final de la lista con NULL
    printf("NULL\n");
}

int main() {
    // Inicializamos la lista vacía (la cabeza apunta a NULL)
    struct Nodo* cabeza = NULL;
    
    printf("--- Iniciando operaciones de la Lista Enlazada ---\n\n");
    
    // Probando la inserción
    insertarAlPrincipio(&cabeza, 10);
    insertarAlPrincipio(&cabeza, 20);
    insertarAlPrincipio(&cabeza, 30);
    insertarAlPrincipio(&cabeza, 40);
    imprimirLista(cabeza);
    printf("\n");
    
    // Probando la búsqueda
    int valorBuscar = 20;
    struct Nodo* resultadoBusqueda = buscar(cabeza, valorBuscar);
    if (resultadoBusqueda != NULL) {
        printf("Busqueda: Se encontro el nodo. Su direccion de memoria es %p y su valor es %d.\n", (void*)resultadoBusqueda, resultadoBusqueda->dato);
    } else {
        printf("Busqueda: Nodo con valor %d no encontrado.\n", valorBuscar);
    }
    printf("\n");
    
    // Probando la modificación
    modificar(cabeza, 30, 35);
    imprimirLista(cabeza);
    printf("\n");
    
    // Probando la eliminación
    eliminar(&cabeza, 40); // Eliminar la cabeza
    imprimirLista(cabeza);
    eliminar(&cabeza, 10); // Eliminar un nodo del final o medio
    imprimirLista(cabeza);
    eliminar(&cabeza, 100); // Intentar eliminar un nodo inexistente
    printf("\n");
    
    return 0;
}
